User registration

Dr1ff opened the pull request.

Dr1ff committed:
Добавил UpdateHandler, CommandProcessor, Паттерн Observer
KeyboardCallbackProcessor

Dr1ff committed:
Добавил UpdateHandler, CommandProcessor, Паттерн Observer
KeyboardCallbackProcessor

Dr1ff committed:
Небольшие изменения процессоров

Dr1ff committed:
Флаги

Dr1ff committed:
исправил корявое описание

Dr1ff committed:
Поправил энтити и скрипт

Dr1ff committed:
Поправил энтити и скрипт

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/StartCommandProcessor.java:
```diff
@@ -0,0 +1,54 @@
+    private final MessageEventPublisher messageEventPublisher;
+    @Override
+    public void process(Update update) {
+//        var user = getUser(update.getMessage().getFrom().getId());
```
забыл удалить?

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/StartCommandProcessor.java:
```diff
@@ -0,0 +1,54 @@
+//            user = userO.get();
+//        }
+//        return user;
+//    }
```
забыл убрать?

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/KeyboardCallbackProcessor.java:
```diff
@@ -0,0 +1,58 @@
+
+@Component
+@RequiredArgsConstructor
+public class KeyboardCallbackProcessor implements CallbackProcessor {
```
название лучше сменить на UserRegionCallbackProcessor или UserCountry...

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/UpdateHandler.java:
```diff
@@ -0,0 +1,37 @@
+           Optional<Command> commandO = Optional.of(Command.valueOf(messageText));
+           commandO.ifPresent(command -> commandProcessors.get(command).process(update));
+       } else if(update.hasCallbackQuery()) {
+           callbackProcessor.process(update.getCallbackQuery());
```
как будет работать обработка коллбеков когда появятся другие процессоры?

Commit:
template of GamesImportTask

Dr1ff committed:
Втащил ObjectMapper
Добавил CallbackCommandData
Добавил команду REGISTER
Добавил зависимости jackson
Добавил Enum Subcommand

Dr1ff committed:
Втащил ObjectMapper
Добавил CallbackCommandData
Добавил команду REGISTER
Добавил зависимости jackson
Добавил Enum Subcommand

Dr1ff committed:
Рефактор всех процессоров

Dr1ff committed:
Написал CountryKeyboardService

Commit:
fix

Dr1ff committed:
Добавил команду BREAK для обработки callbackQuery если пользователь тыкает на пустую кнопку
исправил миграцию таблицы user
исправил анотации entity

Dr1ff committed:
Добавил команду BREAK для обработки callbackQuery если пользователь тыкает на пустую кнопку
исправил миграцию таблицы user
исправил анотации entity

Dr1ff committed:
Клавиатура регионов, обработка колбэка, новые Subcommand

Dr1ff committed:
User registration task

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,167 @@
+    public InlineKeyboardMarkup getMarkup() {
+        var commandData = new CallbackCommandData();
+        commandData.setCommand(Command.REGISTER);
+        setSubcommandData(Subcommand.PREV, commandData, 8);
```
8 вынести в константу

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,167 @@
+    private InlineKeyboardMarkup getNext(CallbackCommandData commandData) {
+        var rows = getRows();
+        int firstIndex = Integer.parseInt(commandData.getSubcommandArgs().get(0));
+        int lastIndex = firstIndex + 9;
```
вынести в константы все числа для которых это возможно

Dr1ff committed:
Keyboards big fix
CallbackDto
CallbackData -> business
CommandArgs,SubcommandArgs -> objects
CommandArgsCreator, SubcommandArgsCreator + impl

Dr1ff committed:
Style fix
Builders for Replays

Dr1ff committed:
Style fix
Builders for Replays

Dr1ff committed:
Debug CallbackParser

Dr1ff committed:
fix

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/handler/UpdateHandler.java:
```diff
@@ -0,0 +1,56 @@
+            Optional<Command> commandO = Optional.of(Command.valueOf(messageText));
+            commandO.ifPresent(command -> commandProcessors.get(command).process(update.getMessage()));
+        } else if (update.hasCallbackQuery()) {
+            var callbackData = callbackParser.fromDto(
+                    objectMapper.readValue(update.getCallbackQuery().getData(), CallbackDto.class));
```
Кажется что нужно это вынести отсюда в отдельный класс, например CallbackDataMapper, и все операции с objectMapper и callbackParser проводить в нем

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    private List<List<InlineKeyboardButton>> getRows() {
+        return List.of(
+                new ArrayList<>(),
+                new ArrayList<>(),
+                new ArrayList<>(),
+                new ArrayList<>()
+        );
```
тут явно цикл или стрим напрашивается

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,198 @@
+    private InlineKeyboardButton getButton(CallbackData callbackData) {
+        return InlineKeyboardButton.builder()
+                .text(callbackData.subcommand().getButtonText())
+                .callbackData(
+                        objectMapper.writeValueAsString(
+                                callbackParser.fromData(callbackData)
```
objectMapper и callbackParser вынести в одно месте, описал выше

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    @SneakyThrows
+    private InlineKeyboardButton getEmptyButton() {
+        return InlineKeyboardButton.builder()
+                .text(" ")
+                .callbackData(
+                        objectMapper.writeValueAsString(callbackParser.fromData(
+                                        CallbackData.builder()
+                                                .command(Command.BREAK)
+                                                .build()
+                                )
+                        )
+                )
+                .build();
+    }
```
Выглядит как переиспользуемая штука, давай вынесем в какой то утилитарный класс

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,198 @@
+    private InlineKeyboardButton getCountryButton(Country country) {
+        return InlineKeyboardButton.builder()
+                .text( /* Страна и ее флаг(emoji) */
+                        EmojiParser.parseToUnicode(country.name())
```
убрать

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/keyboard/CountryKeyboardService.java:
```diff
@@ -0,0 +1,198 @@
+        return InlineKeyboardButton.builder()
+                .text( /* Страна и ее флаг(emoji) */
+                        EmojiParser.parseToUnicode(country.name())
+                                + EmojiManager.getForAlias(country.name().toLowerCase(Locale.ROOT)).getUnicode())
```
давай Locale.ROOT уберем

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/CallbackDto.java:
```diff
@@ -0,0 +1,21 @@
+        @JsonProperty("ca") List<String> commandArgs,
+        @JsonProperty(value = "sa") List<String> subcommandArgs
+) {
+    @Jacksonized
```
убери если не нужна

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/CallbackParser.java:
```diff
@@ -0,0 +1,84 @@
+
+    public CallbackData fromDto(CallbackDto dto) {
+        var builder = new CallbackData.CallbackDataBuilder();
+        if (dto.command() != null) {
```
пометь todo сделать что нибудь с нуллами этими, я потом посмотрю

Rosamonaar replied:
давай для проверки на null везде использовать Objects.isNull()

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/CallbackParser.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public CallbackData fromDto(CallbackDto dto) {
+        var builder = new CallbackData.CallbackDataBuilder();
```
давай собирать объект через конструктор, а не через билдер

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/CallbackData.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Builder;
+
+public record CallbackData(
+        Command command,
+        Subcommand subcommand,
+        CommandArgs commandArgs,
+        SubcommandArgs subcommandArgs
```
нужно все поля которые nullable сделать Optional

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/command/CommandArgsCreator.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface CommandArgsCreator {
+
+    CommandArgs create(List<String> dtoArgs);
+
+    List<String> toList(CommandArgs commandArgs);
```
Более подходящим названием для методов выглядит fromArgsList(List<String>) и toArgsList(CommandArgs)

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/command/CommandArgsCreatorImpl.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public CommandArgs create(List<String> dtoArgs) {
+        return CommandArgsImpl.builder()
```
заменить на конструктор

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/subcommand/CountrySubcommandArgsCreator.java:
```diff
@@ -0,0 +1,36 @@
+public class CountrySubcommandArgsCreator implements SubcommandArgsCreator {
+
+    @Override
+    public SubcommandArgs create(List<String> dtoArgs) {
+        return CountrySubcommandArgs.builder()
+                .country(Country.valueOf(dtoArgs.get(0)))
+                .build();
+    }
+
+    @Override
+    public List<String> toList(SubcommandArgs subcommandArgs) {
+        return List.of(((CountrySubcommandArgs) subcommandArgs).country().name());
```
выше описал нейминг методов

Rosamonaar commented on src/main/java/com/example/nintendoswitchdiscountsbot/service/command/processor/callback/subcommand/SubcommandArgsCreator.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface SubcommandArgsCreator {
+
+    SubcommandArgs create(List<String> dtoArgs);
+
+    List<String> toList(SubcommandArgs subcommandArgs);
```
нейминг

Dr1ff committed:
fix #2

Dr1ff committed:
Resolve conflicts with main

Dr1ff committed:
Resolve conflicts with main#2

Dr1ff committed:
Merge branch 'main' into user_registration

# Conflicts:
#	pom.xml
#	src/main/java/com/example/nintendoswitchdiscountsbot/business/User.java
#	src/main/java/com/example/nintendoswitchdiscountsbot/configuration/ApplicationContextConfiguration.java
#	src/main/java/com/example/nintendoswitchdiscountsbot/entity/GameEntity.java
#	src/main/java/com/example/nintendoswitchdiscountsbot/entity/NotificationEntity.java
#	src/main/java/com/example/nintendoswitchdiscountsbot/entity/UserEntity.java
#	src/main/java/com/example/nintendoswitchdiscountsbot/service/TelegramBot.java
